cloudinit/config/cc_runcmd.py

```diff
@@ -49,6 +49,12 @@
 
     when writing files, do not use /tmp dir as it races with
     systemd-tmpfiles-clean LP: #1707222. Use /run/somedir instead.
+
+.. note::
```

Comment:
Are we sure that this note belongs here?

Our docs have many little notes like this that might be relevant to their adjacent section, but don't serve the majority of users. This [comment I just made on the boothook docs PR](https://github.com/canonical/cloud-init/pull/4650/files#r1427264695) references another example of is problem in our docs. Both that comment and this one point out what seems to me like a smell, and I think we can improve it. Yes, the information you added might be relevant to a user that is reading this section, but a user looking in that section isn't necessarily using snap commands, so it's not relevant to everyone. This note is relevant only to Ubuntu users, and only the subset which want to run snap commands - definitely not to all users of `cc_runcmd`.

This is an example of a standard documentation practice in this project that I think that we can improve on. Many cloud-init docs were written in such a way that relevant information gets grouped together on adjacent pages, however this dilutes the purpose of the original section and leaves users with information that is irrelevant to them, or leaves them with unanswered questions. Writing documentation in this style means that our documentation looks a lot more like a large collection of "notes" than reference material or a user's manual. I think we should prefer to put information in sections that are relevant to all users of that section, rather than having a long tail of implementation details following each section that only maybe applies. The reason for preferring this is that maybe-relevant-notes after sections makes the experience unwieldy to the user. As a user, am I supposed to read all module notes to know how to use the module? If yes, and if we add a note like this to every affected module every time we change something in upstream cloud-init, then I as a user end up reading lots of irrelevant information for my use-case.

I'd like to make the case that the appropriate spot for this would be in our Changelog, maybe in a section titled "breaking changes", or something like that. If we wrote this module from scratch, I don't think that we would have written the `cc_runcmd` documentation like this with an Ubuntu-specific callout. Maybe if we think this requires a dedicated spot in the docs, it could go in a distro-specific page? A snap-specific page? I think that this would increase the signal to noise ratio in our docs over copy-pasting implementation details to all parts of the docs where the details may be relevant.

Same comment to the other docs changes.

Reply:
@s-makin Thoughts on this?

Reply:
As a general principle, I completely agree. I would like us to try and avoid (or at least minimise) the use of notes and asides by finding appropriate "homes" for all these little callouts to live in. I suspect it'll be a bit of a Pandora's box to identify and rehome them all, but probably a worthwhile effort if it helps to streamline the user experience. First step, I think, will be to identify and gather them all so we can gauge the scope of the problem. It might need time on the next roadmap, even if it's a small one.

If there are parts of the docs that *require* distro-specific callouts, it could be worth considering using synced tabs across the docs so that users can select their distro the first time they encounter the tabset, and then only get shown the relevant content by default (instead of ploughing through a list of callouts).

Reply:
This works for me, I've dropped the snap-related  notes in runcmd and and the scripts in general and represented this `BREAKING CHANGE` in the commit message body as well as in the commit summary with the `!`. I think it's worth us adding process in our release note generation to call-out those breaking changes both in github releases and launchpad releases we represent. For discoverability of BREAKING CHANGES in commits, I suggest we also add the trailing bang `!` to the shortened commit summary for related commits/features that do diverge. If agreed, I'll add a PR to [improve our 
pull request template](https://github.com/canonical/cloud-init/blob/main/.github/PULL_REQUEST_TEMPLATE.md) to better document this commit message practice for approaching BREAKING CHANGES.

Reply:
> or discoverability of BREAKING CHANGES in commits, I suggest we also add the trailing bang ! to the shortened commit summary for related commits/features that do diverge

@blackboxsw I meant to comment on this before, but I'm not sure that I agree that we want to require this. It's not intuitive what that means, and while it is in the specification I don't see why that would add anything that BREAKING CHANGE doesn't already accomplish (and more intuitively)